backends/graphics/opengl/pipelines/libretro/parser.cpp

```diff
@@ -0,0 +1,491 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "graphics/opengl/system_headers.h"
+
+#if !USE_FORCED_GLES
+#include "backends/graphics/opengl/pipelines/libretro/parser.h"
+
+#include "common/fs.h"
+#include "common/hash-str.h"
+#include "common/stream.h"
+#include "common/algorithm.h"
+#include "common/tokenizer.h"
+#include "common/ptr.h"
+#include "common/util.h"
+
+#include "common/textconsole.h"
+
+#include <limits.h>
+
+namespace OpenGL {
+namespace LibRetro {
+
+class PresetParser {
+public:
+	ShaderPreset *parseStream(Common::SeekableReadStream &stream);
+
+	const Common::String &getErrorDesc() const { return _errorDesc; }
+private:
+	bool parsePreset(Common::SeekableReadStream &stream);
+	bool lookUpValue(const Common::String &key, Common::String *value);
+	bool lookUpValue(const Common::String &key, bool *value);
+	bool lookUpValue(const Common::String &key, uint *value);
+	bool lookUpValue(const Common::String &key, float *value);
+	bool lookUpValue(const Common::String &key, FilteringMode *value, const FilteringMode defaultValue);
+	bool lookUpValue(const Common::String &key, ScaleType *value, const ScaleType defaultValue);
+	bool lookUpValueScale(const Common::String &key, float *floatValue, uint *uintValue, const ScaleType scaleType);
+
+	template<typename T, typename DefaultT>
+	bool lookUpValue(const Common::String &key, T *value, const DefaultT &defaultValue) {
+		if (_entries.contains(key)) {
+			return lookUpValue(key, value);
+		} else {
+			*value = defaultValue;
+			return true;
+		}
+	}
+
+
+	bool parseTextures();
+	bool parseTexture(const Common::String &id);
+
+	bool parsePasses();
+	bool parsePass(const uint id, const bool isLast);
+	bool parsePassScaleType(const uint id, const bool isLast, ShaderPass *pass);
+	bool parsePassScale(const uint id, ShaderPass *pass);
+
+	typedef Common::HashMap<Common::String, Common::String> StringMap;
+	StringMap _entries;
+
+	Common::String _errorDesc;
+
+	Common::ScopedPtr<ShaderPreset> _shader;
+};
+
+ShaderPreset *PresetParser::parseStream(Common::SeekableReadStream &stream) {
+	_errorDesc.clear();
+	_entries.clear();
+
+	if (!parsePreset(stream)) {
+		return nullptr;
+	}
+
+	_shader.reset(new ShaderPreset);
+	if (!parseTextures()) {
+		return nullptr;
+	}
+
+	if (!parsePasses()) {
+		return nullptr;
+	}
+
+	return _shader.release();
+}
+
+bool PresetParser::parsePreset(Common::SeekableReadStream &stream) {
+	while (!stream.eos()) {
+		Common::String line = stream.readLine();
+		if (stream.err()) {
+			_errorDesc = "Read error";
+			return false;
+		}
+
+		if (line.empty()) {
+			continue;
+		}
+
+		bool empty = true;
+		for (int i = 0; i < line.size(); i++) {
+			if (line[i] == '#')
+				break;
+
+			if (!Common::isSpace(line[i])) {
+				empty = false;
+				break;
+			}
+		}
+
+		if (empty)
+			continue;
+
+		// Split line into key, value pair.
+		// TODO: Files can contain comments starting with '#', we need to
+		// handle this.
```

Comment:
Should this be removed now that we support comments?

Reply:
Done